05-structs.Rmd

```diff
@@ -0,0 +1,467 @@
+    "ref_amb": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.amb",
+    "ref_ann": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.ann",
+    "ref_bwt": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/Homo_sapiens_assembly19.fasta.bwt",
+    "ref_name": "hg19"
+  },
+  "minidata_mutation_calling_v1.dbSNP_vcf_index": "/fh/fast/paguirigan_a/pub/ReferenceDataSets/genome_data/human/hg19/dbsnp_138.b37.vcf.gz.tbi",
+  ...
+}
+```
+
+Now `refGenome` has all the values it needs for our tasks.
+
+In addition, we have replaced all the reference genome inputs in `call BwaMem` with `refGenome`: we can pass information to a task via structs.
+
+Within the `BwaMem` task, we have to to refer to variables inside the struct, such as `refGenome.ref_name` (which has a value of "hg19" using this JSON metadata):
+
+```         
+task BwaMem {
+  input {
+    File input_fastq
+    referenceGenome refGenome
+  }
+  
+  String base_file_name = basename(input_fastq, ".fastq")
+  String ref_fasta_local = basename(refGenome.ref_fasta)
+
+  String read_group_id = "ID:" + base_file_name
+  String sample_name = "SM:" + base_file_name
+  String platform = "illumina"
+  String platform_info = "PL:" + platform   # Create the platform information
+
+
+  command <<<
+    set -eo pipefail
+
+    #can we iterate through a struct??
+    mv ~{refGenome.ref_fasta} .
+    mv ~{refGenome.ref_fasta_index} .
+    mv ~{refGenome.ref_dict} .
+    mv ~{refGenome.ref_amb} .
+    mv ~{refGenome.ref_ann} .
+    mv ~{refGenome.ref_bwt} .
+    mv ~{refGenome.ref_pac} .
+    mv ~{refGenome.ref_sa} .
+
+    bwa mem \
+      -p -v 3 -t 16 -M -R '@RG\t~{read_group_id}\t~{sample_name}\t~{platform_info}' \
+      ~{ref_fasta_local} ~{input_fastq} > ~{base_file_name}.sam 
+    samtools view -1bS -@ 15 -o ~{base_file_name}.aligned.bam ~{base_file_name}.sam
+    samtools sort -@ 15 -o ~{base_file_name}.sorted_query_aligned.bam ~{base_file_name}.aligned.bam
+  >>>
+
+  output {
+    File analysisReadySorted = "~{base_file_name}.sorted_query_aligned.bam"
+  }
+  
+  runtime {
+    memory: "48 GB"
+    cpu: 16
+    docker: "fredhutch/bwa:0.7.17"
+  }
+}
+```
+
+Other tasks in the workflow, such as `ApplyBaseRecalibrator` and `Mutect2TumorOnly` also make use of the reference genome, we so pass `refGenome` to it. For the final task `annovar`, it only requires the reference genome name, and none of the files in the `referenceGenome` struct. We make a stylistic choice to pass only `refGenome.ref_name` to the input of `annovar` task calling, as the task doesn't need the full information of the struct. This stylistic choice is based on the principle of passing on the minimally needed information for a modular piece of code to run, which makes the task `annovar` depend on the minimal amount of inputs.
```

Comment:
Changes to consider:
* "...genome, so we pass `refGenome` to it..."
* "The final task `annovar` only requires..."
* "...to the input of `annovar` task call, as the task..." (whether this noun should be "call" or "calling" is probably a matter of opinion, but I usually see call as the noun and verb)

Also consider adding "This will also save us time and disk space by not having to localize several gigabytes of reference files into the Docker container that `annovar` will be running in." at the very end.